pkg/ocp/actuator.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2019 The OpenShift Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package ocp
+
+import (
+	"context"
+	"fmt"
+	"errors"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	kubernetesErrors "k8s.io/apimachinery/pkg/api/errors"
+	minterv1 "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+)
+
+type OCPActuator struct {
+	Client client.Client
+	Codec  *minterv1.ProviderCodec
+}
+
+const (
+	DataSecretKey = "userData"
+	KubevirtCredentials = "kubevirt-credentials"
+	KubevirtCredentialsNamespace = "kube-system"
+	KubevirtCredentialsSecretKey = "kubeconfig"
+)
+// NewActuator creates a new OCP actuator.
+func NewActuator(client client.Client) (*OCPActuator, error) {
+	codec, err := minterv1.NewCodec()
+	if err != nil {
+		log.WithError(err).Error("error creating OCP codec")
+		return nil, fmt.Errorf("error creating OCP codec: %v", err)
+	}
+
+	return &OCPActuator{
+		Codec:  codec,
+		Client: client,
+	}, nil
+}
+
+// Exists checks if the credentials currently exist.
+// TODO: in the future validate the expiration of the credentials
+func (a *OCPActuator) Exists(ctx context.Context, cr *minterv1.CredentialsRequest) (bool, error) {
+	logger := a.getLogger(cr)
+	logger.Debug("running Exists")
+	var err error
+
+	existingSecret, err := a.getSecret(cr, logger)
+	if err != nil {
+		return false, err
+	}
+
+	return existingSecret != nil, nil
+}
+
+// Create the credentials.
+func (a *OCPActuator) Create(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	logger := a.getLogger(cr)
+	logger.Debug("running create")
+	return a.createSecret(ctx, cr, logger)
+}
+
+// Update the credentials to the provided definition.
+func (a *OCPActuator) Update(ctx context.Context, cr *minterv1.CredentialsRequest) error {
```

Comment:
You do here only create if doesn't exists, what about update existing secret?

Reply:
Why need to update? if the key exist it exist. if no it's not and need to create
As I said - in the future we will have expiration kubeconfig and if the kubeconfig expired exists will return false and then the reconcile will call to create. 
There isn't nothing to update here... 
This is the same behavior as was implemeted in other platforms